File: CPPSolutions/AcWing/DP/LinearDP/Q899_编辑距离.cpp
using namespace std;

#include <iostream>
#include <cstring>

const int M = 1e3 + 10, N = 10 + 3;
char s[M][N];
int f[N][N];
int n, m;

int editDistance(char a[], char b[]) {
    int lenA = strlen(a + 1);
    int lenB = strlen(b + 1);
    for (int i = 0; i <= lenA; i++) {
        f[i][0] = i;
    }
    for (int j = 0; j <= lenB; j++) {
        f[0][j] = j;
    }

    for (int i = 1; i <= lenA; i++) {
        for (int j = 1; j <= lenB; j++) {
            f[i][j] = a[i] == b[j] ? f[i - 1][j - 1] : min(min(f[i - 1][j], f[i][j - 1]), f[i - 1][j - 1]) + 1;
        }
    }

    return f[lenA][lenB];
}

int main() {
    scanf("%d%d", &n, &m);
    for (int i = 0; i < n; i++) {
        scanf("%s", s[i] + 1);
    }
    while (m--) {
        char q[N];
        int limit;
        scanf("%s%d", q + 1, &limit);

        int ans = 0;
        for (int i = 0; i < n; i++) {
            if (editDistance(s[i], q) <= limit) {
                ans++;
            }
        }
        printf("%d\n", ans);
    }
}

File: AcWing/src/DP/LinearDP/Q899_编辑距离.java
package DP.LinearDP;

import java.io.*;
import java.util.StringTokenizer;

/**
 * @author Hyper
 * @date 2022/05/05
 */
public class Q899_编辑距离 {
    public static BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
    public static StringTokenizer tk = new StringTokenizer("");
    public static PrintWriter out = new PrintWriter(new BufferedOutputStream(System.out));
    private static final int N = 10 + 3, M = 1000 + 10;
    private static final char[][] s = new char[M][N];
    private static int[][] f = new int[N][N];

    public static void main(String[] args) throws IOException {
        int n = nextInt(), m = nextInt();
        for (int i = 0; i < n; i++) {
            s[i] = next().toCharArray();
        }

        while (m-- > 0) {
            char[] q = next().toCharArray();
            int limit = nextInt();

            int ans = 0;
            for (int i = 0; i < n; i++) {
                if (editDistance(s[i], q) <= limit) {
                    ans++;
                }
            }
            out.println(ans);
        }

        close();
    }

    private static int editDistance(char[] a, char[] b) {
        int n = a.length, m = b.length;
        for (int i = 0; i <= n; i++) {
            f[i][0] = i;
        }
        for (int j = 0; j <= m; j++) {
            f[0][j] = j;
        }
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (a[i - 1] == b[j - 1]) {
                    f[i][j] = f[i - 1][j - 1];
                } else {
                    f[i][j] = Math.min(Math.min(f[i - 1][j], f[i][j - 1]), f[i - 1][j - 1]) + 1;
                }
            }
        }
        return f[n][m];
    }

    public static int nextInt() throws IOException {
        while (!tk.hasMoreTokens()) {
            tk = new StringTokenizer(in.readLine());
        }
        return Integer.parseInt(tk.nextToken());
    }

    public static String next() throws IOException {
        while (!tk.hasMoreTokens()) {
            tk = new StringTokenizer(in.readLine());
        }
        return tk.nextToken();
    }

    public static void close() throws IOException {
        in.close();
        out.close();
    }
}
